crates/db/types/src/mmr_helpers.rs

```diff
@@ -224,10 +243,330 @@ pub fn find_peak_for_pos(pos: u64, mmr_size: u64) -> DbResult<u64> {
+pub struct MmrMetadata {
+    pub num_leaves: u64,
+    pub mmr_size: u64,
+    pub peak_roots: Vec<Buf32>,
+}
+
+impl MmrMetadata {
+    pub fn empty() -> Self {
+        Self {
+            num_leaves: 0,
+            mmr_size: 0,
+            peak_roots: Vec::new(),
+        }
+    }
+}
+
+/// Result of appending a leaf to the MMR
+#[derive(Debug)]
+pub struct AppendLeafResult {
+    pub leaf_index: u64,
+    pub nodes_to_write: Vec<(u64, [u8; 32])>,
+    pub new_metadata: MmrMetadata,
+}
+
+/// Result of popping a leaf from the MMR
+#[derive(Debug)]
+pub struct PopLeafResult {
+    pub leaf_hash: [u8; 32],
+    pub nodes_to_remove: Vec<u64>,
+    pub new_metadata: MmrMetadata,
+}
+
+/// Pure MMR algorithm logic separated from storage
+///
+/// This struct provides the core MMR algorithms (append, pop) as pure functions
+/// that compute what changes need to be made without actually making them.
+/// Storage implementations apply these changes within their own transaction context.
+#[derive(Debug)]
+pub struct MmrAlgorithm;
+
+impl MmrAlgorithm {
+    /// Compute the result of appending a leaf to the MMR
+    ///
+    /// # Arguments
+    ///
+    /// * `hash` - The hash value to append as a new leaf
+    /// * `metadata` - Current MMR metadata (num_leaves, mmr_size, peak_roots)
+    /// * `get_node` - Closure to read existing nodes (for sibling lookups during merge)
+    ///
+    /// # Returns
+    ///
+    /// `AppendLeafResult` containing:
+    /// - The leaf index of the newly appended leaf
+    /// - All nodes to write (leaf + internal nodes from merging)
+    /// - Updated metadata
+    ///
+    /// # Storage Independence
+    ///
+    /// This function doesn't know about transactions or storage backends.
+    /// The caller provides a closure to read nodes and applies the result
+    /// within their own transaction mechanism.
+    pub fn append_leaf<F, E>(
+        hash: [u8; 32],
+        metadata: &MmrMetadata,
+        get_node: F,
+    ) -> Result<AppendLeafResult, E>
+    where
+        F: Fn(u64) -> Result<[u8; 32], E>,
+        E: From<MmrError>,
+    {
+        let leaf_index = metadata.num_leaves;
+        let leaf_pos = leaf_index_to_pos(leaf_index);
+
+        let mut nodes_to_write = Vec::new();
+
+        // Store the leaf
+        nodes_to_write.push((leaf_pos, hash));
+
+        // Merge along the path to create internal nodes
+        let mut current_pos = leaf_pos;
+        let mut current_hash = hash;
+        let mut current_height = 0u8;
+
+        // Keep merging as long as we have a left sibling
+        loop {
+            let next_pos = current_pos + 1;
+            let next_height = pos_height_in_tree(next_pos);
+
+            // If next position is higher, current is a right sibling - we should merge
+            if next_height > current_height {
+                // Current is right sibling, get left sibling
+                let sibling_position = sibling_pos(current_pos, current_height);
```

Comment:
maybe not worth the optimisation, but since you already know it's a right sibling you can directly get the correct sibling position by `current_pos - sibling_offset(current_height)`.